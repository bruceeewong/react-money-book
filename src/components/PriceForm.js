import React from 'react';
import PropTypes from 'prop-types';

const defaultForm = {
  title: '',
  price: 0,
  date: '',
};

const getPayloadByMode = (mode) => {
  return mode === 'create' ? false : true;
}

class PriceForm extends React.Component {
  constructor(props) {
    super(props);
    let mode = 'create';
    if (this.props.form) {
      mode = 'edit';
    }

    const initForm = this.processForm(this.props.form);

    this.state = {
      mode,
      isInit: true,
      validatePass: false,
      alertMsg: '',
      form: { ...initForm },
    };
  }

  processForm = (form) => {
    if (typeof form !== 'object') {
      return defaultForm;
    }
    let initialForm = {};
    Object.keys(defaultForm).forEach(key => {
      initialForm[key] = form[key];
    });
    return initialForm;
  }

  submitForm = (e) => {
    e.preventDefault();

    if (this.state.isInit) {
      this.setState({
        isInit: false,
      });
    }

    let validatePass, alertMsg;
    this.validateForm(this.state.form, (result, msg) => {
      validatePass = result;
      alertMsg = msg;
    });
    
    this.setState({
      validatePass,
      alertMsg,
    });

    if (!validatePass) return;
    
    this.props.onFormSubmit(this.state.form, getPayloadByMode(this.state.mode));
  }

  validateForm = (form, cb) => {
    let dateReg = /^\d{4}-\d{2}-\d{2}$/;

    let message = '';
    if (!form.title || form.title.trim() === '') {
      message = '标题不能为空';
      cb(false, message);
    } else if (parseInt(form.price) <= 0) {
      message = '价格必须大于0';
      cb(false, message);
    } else if (!form.date) {
      message = '日期不能为空';
      cb(false, message);
    } else if (!dateReg.test(form.date)) {
      message = '日期格式应为 YYYY/mm/dd';
      cb(false, message);
    } else {
      message = '表单校验通过';
      cb(true, message);
    }
  }

  onChange = (e) => {
    let {name, value, type} = e.target;

    if (type === 'number') {
      value = parseInt(value);
    }
    
    this.setState({
      form: {
        ...this.state.form,
        [name]: value,
      },
    });
  }

  render() {
    const {onFormCancel} = this.props;
    
    const {
      isInit, 
      validatePass,
      form,
      alertMsg,
    } = this.state;

    return (
      <section>
        {
          !isInit &&
          !validatePass &&
          <div className="alert alert-danger">{alertMsg}</div>
        }

        <form onSubmit={this.submitForm}>
          <div className="form-group">
            <label htmlFor="form-title">标题 *</label>
            <input 
              id="form-title" 
              name="title" 
              className="form-control"
              value={form.title} 
              onChange={this.onChange}
            />
          </div>
          <div className="form-group">
            <label htmlFor="form-price">金额 *</label>
            <div className="input-group">
              <div className="input-group-prepend">
                <span className="input-group-text">¥</span>
              </div>
              <input 
                id="form-price" 
                name="price" 
                className="form-control"
                value={form.price} 
                type="number" 
                onChange={this.onChange}
              />
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="form-date">日期 *</label>
            <input 
              id="form-date" 
              name="date" 
              className="form-control"
              value={form.date} 
              type="date" 
              onChange={this.onChange}
            />
          </div>
          
          <button type="submit" className="btn btn-primary submit-btn mr-3">提交</button>
          <button className="btn btn-secondary cancel-btn" onClick={onFormCancel}>取消</button>
        </form>
      </section>
    );
  }
}

PriceForm.propTypes = {
  form: PropTypes.object,
  onFormSubmit: PropTypes.func,
  onFormCancel: PropTypes.func,
};

export default PriceForm;
