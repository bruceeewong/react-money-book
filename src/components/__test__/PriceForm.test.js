import React from 'react';
import {mount} from 'enzyme';
import PriceForm from '../PriceForm';

const props = {
  onFormSubmit: jest.fn(),
  onFormCancel: jest.fn(),
};

const fakeForm = {
  title: '购物',
  price: '100',
  date: '2020-10-07',
};

const expectFormNotPass = (wrapper) => {
  expect(wrapper.find('.alert').length).toEqual(1);
  expect(props.onFormSubmit).not.toHaveBeenCalled();
}

const fakeEvent = (name, value) => ({
  target: { 
    name, 
    value,
  },
});

const propsWithForm = {
  form: fakeForm,
  onFormSubmit: jest.fn(),
  onFormCancel: jest.fn(),
};

let wrapper = null;
let wrapperWithForm = null;

describe('test rendering', () => {
  beforeEach(() => {
    wrapper = mount(<PriceForm {...props} />);
    wrapperWithForm = mount(<PriceForm {...propsWithForm} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapperWithForm).toMatchSnapshot();
  });

  it('should render 3 input and 1 form', () => {
    expect(wrapper.find('input').length).toEqual(3);
    expect(wrapper.find('form').length).toEqual(1);
    expect(wrapper.find('.alert').length).toEqual(0);
  });
});

describe('test control default value', () => {
  it('should be empty if no item props', () => {
    wrapper.find('input').forEach(input => {
      const ins = input.instance();
      let val = ins.value;
      if (ins.name === 'price') {
        expect(val).toEqual("0");
      } else {
        expect(val).toEqual('');
      }
    });
  });

  it('should display props value if props form is given', () => {
    expect(wrapperWithForm.find('#form-title').instance().value).toEqual(fakeForm.title);
    expect(wrapperWithForm.find('#form-price').instance().value).toEqual(fakeForm.price);
    expect(wrapperWithForm.find('#form-date').instance().value).toEqual(fakeForm.date);
  });
});

describe('test form validation', () => {
  it('should stop submit and alert if submit empty form', () => {
    wrapper.find('form').simulate('submit');
    expectFormNotPass(wrapper);
  });

  it('should stop submit and alert if submit invalid price', () => {
    let invalidValues = [-1, 0];
    invalidValues.forEach(val => {
      wrapper.find('#form-price').simulate('change', fakeEvent('price', val));
      wrapper.find('form').simulate('submit');
      expectFormNotPass(wrapper);
      expect(wrapper.find('.alert').text()).toEqual('价格必须大于0');
    });
  });

  it('should stop submit and alert if submit invalid date', () => {
    let invalidValues = ['', '2020/10'];

    invalidValues.forEach(date => {
      wrapper.find('#form-date').simulate('change', fakeEvent('date', date));
      wrapper.find('form').simulate('submit');
      expectFormNotPass(wrapper);
    });
  });
});

describe('test create mode', () => {
  it('should be edit mode when pass in prop form', () => {
    expect(wrapper.state('mode')).toEqual('create');
  });

  it('create mode: should submit with [form, false] and no alert if values are valid', () => {
    wrapper.find('#form-title').simulate('change', fakeEvent('title', fakeForm.title));
    wrapper.find('#form-price').simulate('change', fakeEvent('price', fakeForm.price));
    wrapper.find('#form-date').simulate('change', fakeEvent('date', fakeForm.date));

    wrapper.find('form').simulate('submit');
    
    expect(props.onFormSubmit).toHaveBeenCalledWith(fakeForm, false);
  });
});

describe('test edit mode', () => {
  it('should be edit mode when pass in prop form', () => {
    expect(wrapperWithForm.state('mode')).toEqual('edit');
  });

  it('should submit with [form, true] and no alert if values are valid', () => {
    wrapperWithForm.find('#form-title').simulate('change', fakeEvent('title', fakeForm.title));
    wrapperWithForm.find('#form-price').simulate('change', fakeEvent('price', fakeForm.price));
    wrapperWithForm.find('#form-date').simulate('change', fakeEvent('date', fakeForm.date));

    wrapperWithForm.find('form').simulate('submit');
    
    expect(propsWithForm.onFormSubmit).toHaveBeenCalledWith(fakeForm, true);
  });
});

describe('button', () => {
  it('should trigger submit event after click submit button', () => {
    wrapper.find('#form-title').simulate('change', fakeEvent('title', fakeForm.title));
    wrapper.find('#form-price').simulate('change', fakeEvent('price', fakeForm.price));
    wrapper.find('#form-date').simulate('change', fakeEvent('date', fakeForm.date));

    wrapper.find('.submit-btn').simulate('click');
    expect(props.onFormSubmit).toHaveBeenCalledWith(fakeForm, false);
  });

  it('should trigger cancel event after click submit button', () => {
    wrapper.find('.cancel-btn').simulate('click');
    expect(props.onFormCancel).toHaveBeenCalled();
  });
});