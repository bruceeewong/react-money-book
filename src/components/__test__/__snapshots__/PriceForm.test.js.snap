// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test rendering should match snapshot 1`] = `ReactWrapper {}`;

exports[`test rendering should match snapshot 2`] = `ReactWrapper {}`;
